Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,1150 @@
+@@ -966,10 +972,13 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+                 PlayerChunk playerchunk = this.getVisibleChunk(chunkcoordintpair.pair());
+ 
+                 if (playerchunk != null) {
+-                    Chunk chunk = playerchunk.getChunk();
++                    Chunk chunk = playerchunk.getTickingChunk(); //for clarification
+ 
+                     if (chunk != null) {
+                         this.a(entityplayer, apacket, chunk);
++                    } else {
++                        //send unload for non ticking chunks (> 32)
++                        entityplayer.a(chunkcoordintpair); //Paper ticket level pp-view-dist
+                     }
+ 
+                     PacketDebug.a(this.world, chunkcoordintpair);
+@@ -1201,29 +1210,38 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+     }
+ 
+     boolean isOutsideOfRange(ChunkCoordIntPair chunkcoordintpair, boolean reducedRange) {
+-        int chunkRange = world.spigotConfig.mobSpawnRange;
+-        chunkRange = (chunkRange > world.spigotConfig.viewDistance) ? (byte) world.spigotConfig.viewDistance : chunkRange;
+-        chunkRange = (chunkRange > 8) ? 8 : chunkRange;
+-
+-        final int finalChunkRange = chunkRange; // Paper for lambda below
+         //double blockRange = (reducedRange) ? Math.pow(chunkRange << 4, 2) : 16384.0D; // Paper - use from event
+         // Spigot end
+         long i = chunkcoordintpair.pair();
+-
+-        return !this.chunkDistanceManager.d(i) ? true : this.playerMap.a(i).noneMatch((entityplayer) -> {
++        //Paper start ticket level pp-view-dist
++        it.unimi.dsi.fastutil.objects.ObjectSet<EntityPlayer> playerset = this.playerTicketAreaMap.getPlayersTracking(i);
++        if(playerset == null || playerset.isEmpty()) return true;
++        if(!this.chunkDistanceManager.naturalSpawnChunksHave(i)) return true;
++
++        it.unimi.dsi.fastutil.objects.ObjectIterator<EntityPlayer> players = playerset.iterator();
++
++        while(players.hasNext()) {
++            EntityPlayer entityplayer = players.next();
++            int playerChunkRange = world.spigotConfig.mobSpawnRange;
++            playerChunkRange = (playerChunkRange > entityplayer.getTrackViewDistance()) ? (byte) entityplayer.getTrackViewDistance() : playerChunkRange;
++            playerChunkRange = (playerChunkRange > 8) ? 8 : playerChunkRange;
+             // Paper start -
+             com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent event;
+             double blockRange = 16384.0D;
+             if (reducedRange) {
+-                event = new com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent(entityplayer.getBukkitEntity(), (byte) finalChunkRange);
++                event = new com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent(entityplayer.getBukkitEntity(), (byte) playerChunkRange); //Paper ticket level pp-view-dist
```

Comment:
this changes behavior since we're no longer calling it for all players

Reply:
solved in rewrite